src/libmongoc/src/mongoc/mongoc-collection.c

```diff
@@ -825,24 +882,45 @@ mongoc_collection_estimated_document_count (
    }
 
    reply_ptr = reply ? reply : &reply_local;
-   bson_append_utf8 (&cmd, "count", 5, coll->collection, coll->collectionlen);
-
-   ret = _mongoc_client_command_with_opts (coll->client,
-                                           coll->db,
-                                           &cmd,
-                                           MONGOC_CMD_READ,
-                                           opts,
-                                           MONGOC_QUERY_NONE,
-                                           read_prefs,
-                                           coll->read_prefs,
-                                           coll->read_concern,
-                                           coll->write_concern,
-                                           reply_ptr,
-                                           error);
-
-   if (ret) {
-      if (bson_iter_init_find (&iter, reply_ptr, "n")) {
-         count = bson_iter_as_int64 (&iter);
+   if (server_stream->sd->max_wire_version < WIRE_VERSION_4_9) {
+      /* On < 4.9, use actual count command for estimatedDocumentCount */
+      bson_init (&cmd);
+      bson_append_utf8 (&cmd, "count", 5, coll->collection, coll->collectionlen);
+      ret = _mongoc_client_command_with_opts (coll->client,
+                                             coll->db,
+                                             &cmd,
+                                             MONGOC_CMD_READ,
+                                             opts,
+                                             MONGOC_QUERY_NONE,
+                                             read_prefs,
+                                             coll->read_prefs,
+                                             coll->read_concern,
+                                             coll->write_concern,
+                                             reply_ptr,
+                                             error);
+      if (ret) {
+         if (bson_iter_init_find (&iter, reply_ptr, "n")) {
+            count = bson_iter_as_int64 (&iter);
+         }
+      }
+   } else {
+      /* On >= 4.9, use aggregate with collStats for estimatedDocumentCount */
+      _make_aggregate_for_edc (coll, &cmd);
+      ret = mongoc_collection_read_command_with_opts (
+         coll, &cmd, read_prefs, opts, reply_ptr, error);
+
+      if (error && error->code == MONGOC_ERROR_COLLECTION_DOES_NOT_EXIST) {
+         /* Collection does not exist. From spec: return 0 but no err:
+          * https://github.com/mongodb/specifications/blob/master/source/crud/crud.rst#estimateddocumentcount
+          */
+         error = NULL;
```

Comment:
I may be mistaken, but I believe that this will cause memory to leak as we're not resetting error information, but rather setting the pointer to `NULL`. This should either use `bson_set_error` or `memset` to make sure there's no error information.

Reply:
I'm not sure if this a memory leak but I'm sure it's a no-op. Because C passes everything by value, this line changes a copy of the parameter's address and not the actual parameter itself or its contents. If this was passed by reference somehow then this would be a memory leak. Here's a code snippet of what's happening:
```
void faux_mod(int *ptr) {
	ptr = NULL;
	printf("faux_mod: address=%p\n", ptr ? ptr : 0);
}

void real_mod(int *ptr) {
	*ptr = 2;
	printf("real_mod: address=%p,val=%d\n", ptr ? ptr : 0, ptr ? *ptr : 0);
}

int main() {
   int* ptr;
   ptr = (int*) malloc(sizeof(int));
   *ptr = 1;

   printf("address=%p,val=%d\n", ptr ? ptr : 0, ptr ? *ptr : 0);

   faux_mod(ptr);
   printf("address=%p,val=%d\n", ptr ? ptr : 0, ptr ? *ptr : 0);

   real_mod(ptr);
   printf("address=%p,val=%d\n", ptr ? ptr : 0, ptr ? *ptr : 0);

   printf("set to NULL\n");
   ptr = NULL;
   printf("address=%p,val=%d\n", ptr ? ptr : 0, ptr ? *ptr : 0);

   free(ptr);
}

...
# output
address=0x558bcab3e2a0,val=1
faux_mod: address=(nil)
address=0x558bcab3e2a0,val=1
real_mod: address=0x558bcab3e2a0,val=2
address=0x558bcab3e2a0,val=2
set to NULL
address=(nil),val=0

```

Reply:
Yep, you're totally right! Very helpful explanation, thank you @bazile-clyde . I only use `bson_set_error` now.

Reply:
Thanks for that explanation @bazile-clyde, I completely forgot about pass-by-value 🤦‍♂️